tests/unit-tests/test-class-sensei-course-enrolment.php

```diff
@@ -0,0 +1,385 @@
+	 */
+	public function testEnrolmentCheckPositivePrevailsFirst() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Always_Provides::class );
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Never_Provides::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+
+		$this->assertTrue( $course_enrolment->is_enroled( $student_id ), 'One positive provider should always enrol student in course' );
+	}
+
+	/**
+	 * Check for positive result when positive and negative provider exist. Checks for case when positive is registered second.
+	 */
+	public function testEnrolmentCheckPositivePrevailsSecond() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Never_Provides::class );
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Always_Provides::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+
+		$this->assertTrue( $course_enrolment->is_enroled( $student_id ), 'One positive provider should always enrol student in course' );
+	}
+
+	/**
+	 * Check for negative result if no providers that handle this course's registration are handled.
+	 */
+	public function testEnrolmentCheckNoHandlingProviders() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Never_Handles::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+
+		$this->assertFalse( $course_enrolment->is_enroled( $student_id ), 'Student should not be enroled if there are no providers handling enrolment for a course' );
+	}
+
+	/**
+	 * Tests to make sure non-handling providers aren't used. Never_Handles has a true response, but it should never be asked.
+	 */
+	public function testEnrolmentCheckNonHandlingProvidersIgnored() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Never_Handles::class );
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Never_Provides::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+
+		$this->assertFalse( $course_enrolment->is_enroled( $student_id ), 'Non-handling providers should never be used to determine enrolment' );
+	}
+
+	/**
+	 * Tests to make sure non-handling providers aren't used. Never_Handles has a true response, but it should never be asked.
+	 */
+	public function testEnrolmentCheckVersionCachingWorks() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+
+		// Initial request.
+		$this->resetAndSetUpVersiondProvider( false ); // Version : 1; Odd numbers do not provide enrolment.
```

Comment:
Should this say "Versioned"?

Also, we should probably manually set the version to 1 at the beginning. Since it's a static variable, it could technically be changed in another test, which may unexpectedly break this one.

Reply:
68fbd553b0927384b1474dbd9f6bc778514d3a0d